src/lib/OS/SimOS.cc

```diff
@@ -169,72 +164,124 @@ void SimOS::createInitialProcess() {
+        processes_[tid]->context_.sp, 8};
   } else if (Config::get()["Core"]["ISA"].as<std::string>() == "AArch64") {
     // Set the stack pointer register
-    processes_[0]->context_.regFile[arch::aarch64::RegisterType::GENERAL][31] =
-        {processes_[0]->context_.sp, 8};
+    processes_[tid]->context_.regFile[arch::aarch64::RegisterType::GENERAL]
+                                     [31] = {processes_[tid]->context_.sp, 8};
     // Set the system registers
     // Temporary: state that DCZ can support clearing 64 bytes at a time,
     // but is disabled due to bit 4 being set
-    processes_[0]->context_.regFile[arch::aarch64::RegisterType::SYSTEM]
-                                   [arch->getSystemRegisterTag(
-                                       ARM64_SYSREG_DCZID_EL0)] = {
+    processes_[tid]->context_.regFile[arch::aarch64::RegisterType::SYSTEM]
+                                     [arch->getSystemRegisterTag(
+                                         ARM64_SYSREG_DCZID_EL0)] = {
         static_cast<uint64_t>(0b10100), 8};
   }
 
-  // In a simulation's initial state, all cores will be idle. Only
-  // 'scheduled' processes may be sent to a core for execution therefore
-  // we must update the initial processes status and push it to the
-  // scheduledProcs queue
-  processes_[0]->status_ = procStatus::scheduled;
-  scheduledProcs_.push(processes_[0]);
+  // If this is the initial process (tid = 0) then add to scheduledProcs_ as
+  // only processes in scheduledProcs_ can be scheduled onto an idle core (all
+  // cores begin in an idle state).
+  // Otherwise, add the new process to waitingProcs_.
+  if (tid == 0) {
+    processes_[tid]->status_ = procStatus::scheduled;
+    scheduledProcs_.push(processes_[tid]);
+  } else {
+    processes_[tid]->status_ = procStatus::waiting;
+    waitingProcs_.push(processes_[tid]);
+  }
+
+  return tid;
 }
 
-void SimOS::createSpecialFileDirectory() const {
-  simeng::SpecialFileDirGen SFdir = simeng::SpecialFileDirGen();
-  // Remove any current special files dir
-  SFdir.RemoveExistingSFDir();
-  // Create new special files dir
-  SFdir.GenerateSFDir();
+const std::shared_ptr<Process>& SimOS::getProcess(uint64_t tid) {
+  auto proc = processes_.find(tid);
+  if (proc == processes_.end()) {
+    // If TID doesn't exist then hard exit
+    std::cerr << "[SimEng:SimOS] ERROR : Process with TID `" << tid
+              << "` does not exist.\n";
+    exit(1);
+  }
+  return proc->second;
+}
+
+void SimOS::terminateThread(uint64_t tid) {
+  auto proc = processes_.find(tid);
+  if (proc == processes_.end()) {
+    // If process with TID doesn't exist, return early
+    return;
+  }
+  switch (proc->second->status_) {
+    case procStatus::executing:
+      // Set core's status to idle
+      for (auto core : cores_) {
+        if (core->getCurrentTID() == tid) {
+          core->setStatus(CoreStatus::idle);
+          break;
+        }
+      }
+      break;
+    case procStatus::scheduled:
```

Comment:
Why can't the removal from scheduledProcs_ or waitingProcs_ take place here, given that we are terminating the process, wouldn't it be easier to just do it here?

Reply:
you can use std::deque if std::queue doesn't allow the functionality

Reply:
changed to `std::deque` so processes can be removed immidatley

Reply:
Having tried an implementation of deque, I think the current implementation is better. 

The removal logic required is :
```
      auto schedItr =
          std::find_if(scheduledProcs_.begin(), scheduledProcs_.end(),
                       [tid](const std::shared_ptr<Process>& proc) {
                         return tid == proc->getTID();
                       });
      scheduledProcs_.erase(schedItr);
```
Which isn't as nice in my opinion, and is most likely less performant.

Let me know your thoughts

Reply:
Resolving it as we discussed it offline.